sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -2358,6 +2360,74 @@ class DataFrameSuite extends QueryTest
     val df = Seq((1, new CalendarInterval(1, 2, 3))).toDF("a", "b")
     checkAnswer(df.selectExpr("b"), Row(new CalendarInterval(1, 2, 3)))
   }
+
+  test("array encoder with different types") {
```

Comment:
Shall we add `SPARK-31552: ` prefix? This is a bug for all old Spark versions. For backporting to branch-2.4, it would be helpful.